packages/core/src/lib/filter/subscription.ts

```diff
@@ -0,0 +1,286 @@
+import { Stream } from "@libp2p/interface/connection";
+import {
+  CustomEvent,
+  EventEmitter,
+  EventHandler
+} from "@libp2p/interface/events";
+import type { Peer } from "@libp2p/interface/peer-store";
+import type {
+  ContentTopic,
+  IDecodedMessage,
+  IDecoder,
+  IProtoMessage,
+  PubSubTopic,
+  SubscriptionEventMap
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import { groupByContentTopic, Logger } from "@waku/utils";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { FilterSubscribeResponse, FilterSubscribeRpc } from "./filter_rpc.js";
+
+const log = new Logger("waku:filter-v2:subscription");
+
+// one subscription object per pubsub topic
+export class Subscription extends EventEmitter<SubscriptionEventMap> {
+  private readonly peer: Peer;
+  private readonly pubsubTopic: PubSubTopic;
+  private newStream: (peer: Peer) => Promise<Stream>;
+  private readonly decoders: Map<ContentTopic, IDecoder<IDecodedMessage>>;
+
+  constructor(
+    pubsubTopic: PubSubTopic,
+    remotePeer: Peer,
+    newStream: (peer: Peer) => Promise<Stream>,
+    decoders: IDecoder<IDecodedMessage>[]
+  ) {
+    super();
+    this.peer = remotePeer;
+    this.pubsubTopic = pubsubTopic;
+    this.newStream = newStream;
+    this.decoders = new Map(
+      decoders.map((decoder) => [decoder.contentTopic, decoder])
```

Comment:
If I understand it correctly - it means there is only one decoder per content topic. 
I think we should support many decoders per content topic as it is done in Relay.

Reply:
It keeps this construct extendable later when we can introduce other `Decoders`. 
Even now someone could try to subscribe with regular decoder and the one from `message-encryption` or build their own.